internal/resources/aws/sns_topic.go

```diff
@@ -1,62 +1,230 @@
+	Address                 string
+	Region                  string
+	RequestSizeKB           *float64 `infracost_usage:"request_size_kb"`
+	MonthlyRequests         *int64   `infracost_usage:"monthly_requests"`
+	HTTPSubscriptions       *int64   `infracost_usage:"http_subscriptions"`
+	EmailSubscriptions      *int64   `infracost_usage:"email_subscriptions"`
+	KinesisSubscriptions    *int64   `infracost_usage:"kinesis_subscriptions"`
+	MobilePushSubscriptions *int64   `infracost_usage:"mobile_push_subscriptions"`
+	MacOSSubscriptions      *int64   `infracost_usage:"macos_subscriptions"`
+	SMSSubscriptions        *int64   `infracost_usage:"sms_subscriptions"`
+	SMSNotificationPrice    *float64 `infracost_usage:"sms_notification_price"`
 }
 
 var SNSTopicUsageSchema = []*schema.UsageItem{
 	{Key: "request_size_kb", ValueType: schema.Float64, DefaultValue: 0},
 	{Key: "monthly_requests", ValueType: schema.Int64, DefaultValue: 0},
+	{Key: "http_subscriptions", ValueType: schema.Int64, DefaultValue: 0},
+	{Key: "email_subscriptions", ValueType: schema.Int64, DefaultValue: 0},
+	{Key: "kinesis_subscriptions", ValueType: schema.Int64, DefaultValue: 0},
+	{Key: "mobile_push_subscriptions", ValueType: schema.Int64, DefaultValue: 0},
+	{Key: "macos_subscriptions", ValueType: schema.Int64, DefaultValue: 0},
+	{Key: "sms_subscriptions", ValueType: schema.Int64, DefaultValue: 0},
+	{Key: "sms_notification_price", ValueType: schema.Float64, DefaultValue: 0.0075},
+}
+
+// This is an experiment to see if using an explicit structure to define the cost components
+// can enable anything interesting (e.g. list what cost components could apply to a resource
+// without having any IaAC)
+func (r *SNSTopic) CostComponents() []*schema.CostComponent {
+	return []*schema.CostComponent{
+		r.APIRequestsCostComponent(nil),
+		r.HTTPNotificationsCostComponent(nil, nil),
+		r.EmailNotificationsCostComponent(nil, nil),
+		r.KinesisNotificationsCostComponent(nil, nil),
+		r.MobilePushNotificationsCostComponent(nil, nil),
+		r.MacOSNotificationsCostComponent(nil, nil),
+		r.SMSNotificationsCostComponent(nil, nil, nil),
+	}
+}
+
+func (r *SNSTopic) APIRequestsCostComponent(requests *int64) *schema.CostComponent {
+	var q *decimal.Decimal
+	if requests != nil {
+		if *requests > 1000000 {
+			q = decimalPtr(decimal.NewFromInt(*requests - 1000000))
+		} else {
+			q = &decimal.Zero
+		}
+	}
+	return &schema.CostComponent{
+		Name:            "API requests (over 1M)",
+		Unit:            "1M requests",
+		UnitMultiplier:  decimal.NewFromInt(1000000),
+		MonthlyQuantity: q,
+		ProductFilter: &schema.ProductFilter{
+			VendorName:    strPtr("aws"),
+			Region:        strPtr(r.Region),
+			Service:       strPtr("AmazonSNS"),
+			ProductFamily: strPtr("API Request"),
+		},
+		PriceFilter: &schema.PriceFilter{
+			StartUsageAmount: strPtr("1000000"),
+		},
+	}
+}
+
+func (r *SNSTopic) HTTPNotificationsCostComponent(subscriptions, requests *int64) *schema.CostComponent {
```

Comment:
are these notification cost component methods exported for a reason? If not let's alter them to be un-exported.